Astropy does not build on MacOSX with Xcode 12.0.0

<!-- This comments are hidden when you submit the issue,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- Please have a search on our GitHub repository to see if a similar
issue has already been posted.
If a similar issue is closed, have a quick look to see if you are satisfied
by the resolution.
If not please go ahead and open an issue! -->

<!-- Please check that the development version still produces the same bug.
You can install development version with
pip install git+https://github.com/astropy/astropy
command. -->

### Description
<!-- Provide a general description of the bug. -->

When I try to build astropy on my Mac (Catalina + Xcode 12.0) it fails with the following error:

```
  cextern/cfitsio/lib/group.c:5664:8: error: implicit declaration of function 'getcwd' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    if (!getcwd(buff,FLEN_FILENAME))
         ^
  1 warning and 1 error generated.
  error: command 'gcc' failed with exit status 1

(astropy) ➜  astropy-tmp git:(master) gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 12.0.0 (clang-1200.0.32.2)
Target: x86_64-apple-darwin19.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Xcode 12.0 was just released Sep. 16 and I'm guessing that it was automatically updated on my Mac since I did not request an update. Astropy is apparently not alone, e.g. https://gitlab.com/graphviz/graphviz/-/issues/1826.

There is a long slack thread with helpful inputs from @saimn, @dhomeier and @manodeep here: https://astropy.slack.com/archives/C067V74GK/p1600984862007700

A workaround is:
```
CFLAGS=-Wno-error=implicit-function-declaration pip install -e .
```

### Expected behavior
<!-- What did you expect to happen. -->

Astropy builds and runs tests.

### Actual behavior
<!-- What actually happened. -->
<!-- Was the output confusing or poorly described? -->

Build failed as shown.

### Steps to Reproduce
<!-- Ideally a code example could be provided so we can run it ourselves. -->
<!-- If you are pasting code, use triple backticks (```) around
your code snippet. -->
<!-- If necessary, sanitize your screen output to be pasted so you do not
reveal secrets like tokens and passwords. -->

1. Checkout astropy at master, `git clean -fxd`
2. `tox -e test` or `pip install -e .`

### System Details
<!-- Even if you do not think this is necessary, it is useful information for the maintainers.
Please run the following snippet and paste the output below:
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import numpy; print("Numpy", numpy.__version__)
import astropy; print("astropy", astropy.__version__)
import scipy; print("Scipy", scipy.__version__)
import matplotlib; print("Matplotlib", matplotlib.__version__)
-->
```
Darwin-19.5.0-x86_64-i386-64bit
Python 3.7.7 (default, Mar 26 2020, 10:32:53) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.18.1
astropy 4.2.dev715+gfaccb8b41
Scipy 1.4.1
Matplotlib 3.1.3
```

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -1027,6 +1027,8 @@ astropy.io.fits
 - Fix repr for commentary cards and strip spaces for commentary keywords.
   [#10640]
 
+- Fix compilation of cfitsio with Xcode 12. [#10772]
+
 astropy.io.misc
 ^^^^^^^^^^^^^^^
 

--- a/astropy/io/fits/setup_package.py
+++ b/astropy/io/fits/setup_package.py
@@ -42,6 +42,8 @@ def _get_compression_extension():
                 '-Wno-declaration-after-statement'
             ])
 
+            cfg['define_macros'].append(('HAVE_UNISTD_H', None))
+
             if not debug:
                 # these switches are to silence warnings from compiling CFITSIO
                 # For full silencing, some are added that only are used in

